### DISCUSSION TITLE
Custom Sentencer causes poor ner training performance

### DISCUSSION THREAD

**Discussion by micmizer** [explosion/spaCy#12873] [Help: Model Advice] [answered]
Seeing some odd behavior with NER and a custom sentencer. The sentencer looks like this and basically sets the next token after \r or \n or any combination of \r\n, \n, \n\n, etc.

```
@Language.component("custom_boundaries")
def sentence_splitter(doc):
    delimiter_pattern = re.compile(r"(\r?\n\s*)+|(\n\s*)+")
    for i, token in enumerate(doc):
        if delimiter_pattern.fullmatch(token.text):
            doc[i + 1].is_sent_start = True
    
    for token in doc:
        if not token.is_sent_start:
            token.is_sent_start = False
    
    return doc
```

Using a very basic pipeline and config generated from the quickstart widget.

pipeline = ["tok2vec","custom_boundaries","ner"]

[components.custom_boundaries]
factory = "custom_boundaries"

Without custom_boundaries NER F1 score is 34.55.

```
E    #       LOSS TOK2VEC  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  ------------  --------  ------  ------  ------  ------
  0       0          0.00    131.86    0.00    0.00    0.00    0.00
...
...
 21    5800      47232.62    605.52   34.55   30.77   39.40    0.35
```

With the custom_boundaries component the F1 score is horrible in comparison at 0.45:

```
E    #       LOSS TOK2VEC  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  ------------  --------  ------  ------  ------  ------
  0       0          0.00    131.86    0.00    0.00    0.00    0.00
...
...
  8    3400     425175.18   1953.12    0.45    0.36    0.60    0.00
```

Here is an example of what our sentence structure should look like in comparison to the default behavior.

```
Default: ['some sentence.\r\n\r\nAnother sentence\n\nA', 'third sentence\n\nA fourth\nand a fifth\n\n\nthe end.']

custom_boundaries: ['some sentence.\r\n\r\n', 'Another sentence\n\n', 'A third sentence\n\n', 'A fourth\n', 'and a fifth\n\n\n', 'the end.']

Tokenized sentences:

['some', 'sentence.', '\r\n\r\n']

['Another', 'sentence', '\n\n']

['A', 'third', 'sentence', '\n\n']

['A', 'fourth', '\n']

['and', 'a', 'fifth', '\n\n\n']

['the', 'end.']
```

I find this quite confusing as our entities traverse multiple sentences/lines. I would not have expected NER to perform much better when the sentence boundaries are wrong?

**Comment by adrianeboyd** [comment, marked as answer]
The `ner` component has been developed for traditional named entities, which are typically short noun phrases that never cross sentence boundaries. There's a hard-coded constraint in the `ner` component to not predict any entities across sentence boundaries.

Your spans don't sound like named entities, so `ner` might not be the best choice, but if it's working fine otherwise, then a simple solution is to reorder the pipeline components so that the sentence boundaries are set after `ner`. But you might also want to consider testing other components like `spancat` that are more flexible in terms of handling longer spans that don't look like short noun phrases.

**Reply by micmizer** [reply]
Thank you. I have converted my "entities" to "spans" using something like the following. Is there documentation for how to create spancat annotations? The only thing i could find is other threads on this discussion board, from which i have adapted the following:

```
for text, annotations in train_data:  # validation_data train_data
    ann_entities = annotations.get('entities', [])
    ann_spans = annotations.get('spans', [])

    doc = nlp.make_doc(text)

    if ann_entities:
        ents = []
        for start, end, label in ann_entities:
            span = doc.char_span(start, end, label=label, alignment_mode="strict")
            if span:
                ents.append(span)
        doc.ents = ents

    if ann_spans:
        doc.spans['sc'] = []
        for start, end, label in ann_spans:
            span = doc.char_span(start, end, label=label, alignment_mode="strict")
            if span:
                #print(f"Creating span: Start: {start}, End: {end}, Label: {label}, Text: {repr(text[start:end])}")
                doc.spans["sc"].append(span)
            else:
                print(f"Error creating span: Start: {start}, End: {end}, Label: {label}, Text: {text[start:end]}")

    doc_bin.add(doc)
```

Using a very basic config.cfg from the quickstart:

```
[paths]
train = "./corpus/train_data.spacy"
dev = "./corpus/dev_data.spacy"
vectors = "en_core_web_lg"
init_tok2vec = null
raw_text = "./corpus/corpus.txt"

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","spancat"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.spancat]
factory = "spancat"
max_positive = null
scorer = {"@scorers":"spacy.spancat_scorer.v1"}
spans_key = "sc"
threshold = 0.5

[components.spancat.model]
@architectures = "spacy.SpanCategorizer.v1"

[components.spancat.model.reducer]
@layers = "spacy.mean_max_reducer.v1"
hidden_size = 128

[components.spancat.model.scorer]
@layers = "spacy.LinearLogistic.v1"
nO = null
nI = null

[components.spancat.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.spancat.suggester]
@misc = "spacy.ngram_suggester.v1"
sizes = [1,2,3]

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,1000,2500,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 256
depth = 8
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
annotating_components = []
before_to_disk = null
before_update = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
spans_sc_f = 1.0
spans_sc_p = 0.0
spans_sc_r = 0.0

[pretraining]

[initialize]
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```


Spancat dosent seem to be training as it lists all 0's. I found some other threads which mention this but they do not have any real resolution. Is this just a bug? Spancat does appear to work with the resulting model however it is surprisingly very poor compared to training as NER. Am i doing something incorrect here? I suspected my training data but i do not get any errors when creating it and after inspecting the created spans they look correct.


```
=========================== Initializing pipeline ===========================
[2023-07-31 13:36:27,441] [INFO] Set up nlp object from config
[2023-07-31 13:36:27,457] [INFO] Pipeline: ['tok2vec', 'spancat']
[2023-07-31 13:36:27,460] [INFO] Created vocabulary
[2023-07-31 13:36:29,069] [INFO] Added vectors: en_core_web_lg
[2023-07-31 13:36:29,070] [INFO] Finished initializing nlp object
[2023-07-31 13:36:33,469] [INFO] Initialized pipeline components: ['tok2vec', 'spancat']
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['tok2vec', 'spancat']
ℹ Initial learn rate: 0.001
E    #       LOSS TOK2VEC  LOSS SPANCAT  SPANS_SC_F  SPANS_SC_P  SPANS_SC_R  SCORE 
---  ------  ------------  ------------  ----------  ----------  ----------  ------
  0       0         38.47       1551.49        0.00        0.00        0.30    0.00
  0     200          5.58        249.60        0.00        0.00        0.00    0.00
  0     400          0.00          3.01        0.00        0.00        0.00    0.00
  0     600          0.00          8.01        0.00        0.00        0.00    0.00
  1     800          0.00          3.01        0.00        0.00        0.00    0.00
  1    1000          0.00          2.00        0.00        0.00        0.00    0.00
  1    1200          0.00          9.00        0.00        0.00        0.00    0.00
  2    1400          0.00          6.00        0.00        0.00        0.00    0.00
  2    1600          0.00          4.00        0.00        0.00        0.00    0.00
✔ Saved pipeline to output directory
```

**Reply by adrianeboyd** [reply]
The issue is probably the default spancat `suggester`, which is currently set to 1-3-grams, so the model isn't predicting any spans longer than 3 tokens, and it sounds like you have longer spans.

The `spancat` component is very flexible, but also needs a bit more task-specific configuration than `ner` when you start out. You can use `spacy debug data` to see a summary of your spans annotation.

If you have lots of long spans, the n-gram suggester is probably not going to work well (you'll run out of RAM), so you might want to look at using `span_finder` or another type of suggester. Some more info: https://spacy.io/api/spancategorizer#suggesters

Again, if `ner` is working well for your data, you can just move the sentence boundary component later in the pipeline.

**Reply by micmizer** [reply]
NER is working well considering the poor training scores and the fact that our entities traverse many lines. I suspect the leading contributor to that is probably our entity_ruler. It is still however not performing as well as we would like. Especially considering how our patterns are pretty consistent and semi-structured. 

When running debug data i get the following span categorization:

```
============================ Span Categorization ============================

Spans Key   Labels                        
---------   ------------------------------
sc          {'EMAIL_SIG', 'LOGS', 'HEADERS'}

ℹ Span characteristics for spans_key 'sc'
ℹ SD = Span Distinctiveness, BD = Boundary Distinctiveness

Span Type      Length     SD     BD     N
------------   ------   ----   ----   ---
EMAIL_SIG       36.49   0.53   3.39   360
LOGS            67.16   1.19   3.62    70
HEADERS         80.67   1.09   3.16   153
------------   ------   ----   ----   ---
Wgt. Average    51.77   0.76   3.36     -

ℹ Over 90% of spans have lengths of 1 -- 393 (min=1, max=2251). The
most common span lengths are: 1 (0.34%), 2 (0.51%), 3 (1.2%), 4 (1.89%), 5
(1.89%), 6 (1.72%), 7 (1.03%), 8 (0.34%), 9 (0.86%), 10 (1.37%), 12 (0.86%), 13
(1.37%), 14 (0.69%), 15 (2.06%), 16 (1.2%), 17 (2.23%), 18 (0.69%), 19 (1.72%),
20 (0.69%), 21 (0.51%), 22 (0.86%), 23 (0.51%), 24 (0.51%), 25 (0.86%), 26
(0.51%), 27 (0.86%), 28 (1.03%), 29 (0.34%), 30 (0.86%), 31 (0.86%), 32 (1.03%),
33 (0.51%), 34 (0.69%), 35 (0.51%), 36 (0.69%), 37 (0.86%), 38 (0.51%), 39
(0.34%), 40 (0.34%), 42 (0.34%), 43 (0.34%), 44 (0.34%), 45 (0.51%), 46 (0.51%),
48 (1.03%), 49 (0.69%), 50 (1.37%), 51 (0.51%), 52 (0.51%), 53 (0.51%), 55
(0.69%), 56 (1.03%), 57 (0.69%), 58 (0.86%), 59 (0.34%), 60 (1.03%), 61 (0.34%),
62 (1.54%), 63 (0.86%), 64 (0.51%), 65 (1.03%), 66 (1.2%), 67 (0.86%), 68
(0.34%), 69 (1.2%), 71 (0.86%), 72 (0.51%), 73 (1.2%), 75 (0.51%), 76 (0.86%),
77 (0.34%), 78 (1.37%), 79 (1.2%), 80 (1.03%), 81 (0.34%), 83 (0.34%), 84
(0.86%), 86 (0.51%), 87 (0.69%), 89 (1.2%), 91 (0.34%), 92 (0.69%), 93 (1.37%),
94 (0.51%), 95 (1.2%), 96 (0.69%), 98 (0.86%), 99 (1.03%), 100 (0.34%), 101
(0.51%), 104 (0.86%), 105 (0.34%), 106 (0.51%), 107 (0.51%), 112 (0.51%), 113
(0.34%), 114 (0.34%), 115 (0.34%), 118 (0.69%), 119 (0.34%), 121 (0.34%), 123
(0.34%), 124 (0.69%), 125 (0.34%), 129 (0.69%), 131 (0.51%), 132 (0.51%), 136
(0.51%), 140 (0.51%), 141 (0.86%), 142 (0.34%), 146 (0.34%), 150 (0.34%), 152
(0.34%), 154 (0.51%), 162 (0.34%), 164 (0.34%), 172 (0.34%), 180 (0.34%), 204
(0.34%), 207 (0.34%), 222 (0.34%), 233 (0.34%), 302 (0.34%), 311 (0.34%), 393
(0.34%). If you are using the n-gram suggester, note that omitting infrequent
n-gram lengths can greatly improve speed and memory usage.
⚠ Spans may not be distinct from the rest of the corpus
✔ Boundary tokens are distinct from the rest of the corpus
✔ Good amount of examples for all labels
✔ Examples without occurrences available for all labels
```

I tested multiple different size variations along with sizes = [37, 70, 81], and different variations of ngram_suggester and ngram_range_suggester. The results are always the same, according to the training pipeline scores spancat will not train.

Per your suggestion I have also tried span_finder + spancat_singlelabel (with preset_spans_suggester) but was also unable to get it to train. 

We have a pretty small set of train and validation data. Eventually i joined the two together and was able to get the span_finder + spancat_singlelabel training.

It appears spancat needs more training data then NER? Spancat also appears to have better predictions albeit less frequent compared to NER. We will have to produce some more training data and test/tweak.

Thanks for your help.

**Reply by adrianeboyd** [reply]
Yeah, such long spans are going to be tricky, and guessing from the labels, this is also a bit outside of what many NLP models are designed to handle (just not natural language exactly). Even with more data, I wouldn't expect to get good performance from `ner` for these kinds of spans.

And you're right that this isn't a lot of training data. For data with non-overlapping named entities and not other types of spans, `ner` will nearly always outperform `spancat` for the same amount of training data because it's designed for this exact task.

Maybe it would be easier if you restructured your task to be a line-by-line classification of the text? Then the suggester can suggest lines and `spancat` is only categorizing them? (This would be similar conceptually to running `textcat` on each line individually.)

With your original spans and `spancat`, I don't think the n-gram suggester is going to work at all for this, so your best options are using `span_finder` and/or `span_ruler`. It sounds like you might be able to identify a large percentage of the spans with regex patterns?

The configuration for `span_finder` can be a bit tricky, make sure you have it added to `[training.annotating_components]`. If it's helpful, here's a demo project that compares different types of suggesters including `span_finder` with example configs:

https://github.com/adrianeboyd/workshop-dh2023/tree/main/litbank

**Reply by micmizer** [reply]
> Yeah, such long spans are going to be tricky, and guessing from the labels, this is also a bit outside of what many NLP models are designed to handle (just not natural language exactly).

Our corpus is pretty challenging in that regard. Typically 25% of a given doc is junk we want to remove. Sometimes 90%. It makes it very challenging. Regular regex can typically get all of the headers but LOGS and EMAIL_SIG it usually just ends up leaving fragments of words and/or characters. The goal with NER or spancat would be to remove the entire line (sentence) for all three labels. 

> Maybe it would be easier if you restructured your task to be a line-by-line classification of the text? Then the suggester can suggest lines and spancat is only categorizing them? (This would be similar conceptually to running textcat on each line individually.)

Are you suggesting tweaking my annotations to be a single sentence instead of the entire doc and then creating an individual doc for each sentence and allowing spancat to make predictions that way?

> It sounds like you might be able to identify a large percentage of the spans with regex patterns?

For headers and email_sig a good portion is already, or can be caught by the entity/span ruler.

**Reply by adrianeboyd** [reply]
With `textcat` you would need to create individual docs for each sentence (and lose the surrounding context), so I would suggest using a suggester that suggests lines (through sentences if that's easy for you; you can copy the sentence suggester from `spacy-experimental` if you like -- I wouldn't use the `spacy-experimental` package, just copy the sentence suggester as custom code) and having `spancat` only label each line.

Convert your annotation to be by line instead of for long spans, so:

```
LOG[a b c d 
a b c d
a b d e]
text text text
SIG[a a a a]
```

->

```
LOG[a b c d]
LOG[a b c d]
LOG[a b c e]
text text text 
SIG[a a a a]
```

If the model has to identify the start token, end token, and the label for potentially long spans, this is a pretty hard task.

If you convert this to a line-labeling task, you provide the start and end tokens with a rule-based component and all the statistical model has to do is label each line, so the problem is much easier and I think you're much more likely to see usable performance.

You do have to be really careful that your annotated spans match your suggested spans, but if you're splitting on newlines this should be relatively easy to do.

**Reply by micmizer** [reply]
Since our patterns encompass the entire line for multiple lines i was able to easily split the annotations into 1 per line as you suggested above. This had a tremendous impact on NER. F1 score for example going from 14 to 80~. Pretty happy with those results and the actual predictions it is making.

I have been having some problems training spancat and even just span_finder by itself. Not sure exactly what the problem was but it seems to have resolved itself after re-generating my training data. Something was silently failing. 

Side question. Should spans ruler, span_finder and  spancat always have the same spans_key? I have only tested with them all set to "sc". but when i train a pipeline like: ['tok2vec', 'span_ruler', 'span_finder'] i only get training scores for spans_sc_x and nothing for span_finder?

```
E    #       LOSS TOK2VEC  LOSS SPAN_...  SPANS_RULER_F  SPANS_RULER_P  SPANS_RULER_R  SPANS_SC_F  SPANS_SC_P  SPANS_SC_R  SPANS_FIND...  SPANS_FIND...  SPANS_FIND...  SCORE 
---  ------  ------------  -------------  -------------  -------------  -------------  ----------  ----------  ----------  -------------  -------------  -------------  ------
 28   12600          3.35         772.54           0.00           0.00           0.00       51.84       36.64       88.59           0.00           0.00           0.00    0.09
 29   12800          3.13         699.62           0.00           0.00           0.00       51.86       36.62       88.83           0.00           0.00           0.00    0.09
 30   13000          3.45         666.09           0.00           0.00           0.00       51.94       36.65       89.10           0.00           0.00           0.00    0.09
 30   13200          3.42         725.44           0.00           0.00           0.00       51.75       36.53       88.67           0.00           0.00           0.00    0.09
```

> so I would suggest using a suggester that suggests lines (through sentences if that's easy for you; you can copy the sentence suggester from spacy-experimental if you like -- I wouldn't use the spacy-experimental package, just copy the sentence suggester as custom code) and having spancat only label each line.

I have only used preset_spans_suggester in my prior attempts. Are you referring to sentence_suggester function [here](https://github.com/explosion/spacy-experimental/blob/master/spacy_experimental/span_suggesters/sentence_suggester.py)

**Reply by adrianeboyd** [reply]
`preset_spans_suggester` is easier than dealing with the sentence suggester, so that's fine.

If it doesn't seem to be training or you're seeing all 0s, some of the main places to check:

* the evaluation requires `span_ruler` or `span_finder` to have the same `spans_key` as in the training data
* double-check the `spans_key` settings used for `spancat` including the suggester
* make sure that the components that are required to provide spans for `spancat` are all in `annotating_components`

Be aware that the spans set by `span_ruler` and `span_finder` wouldn't be combined automatically. You might be accidentally overwriting spans at some point in your pipeline. `spancat` also overwrites the `spans_key` spans.

If you want to use `span_finder` + `span_ruler` for preset spans, add `span_ruler` after `span_finder` and set `overwrite=False`.
